applications/ShapeOptimizationApplication/custom_utilities/mapping/mapper_vertex_morphing_symmetric.cpp

```diff
@@ -0,0 +1,363 @@
+    } else if (mMapperSettings["revolution"].GetBool()) {
+        mpSymmetry = Kratos::make_unique<SymmetryRevolution>(mrOriginModelPart, mrDestinationModelPart, mMapperSettings["revolution_settings"]);
+    } else {
+        KRATOS_ERROR << "No symmetry type specified" << std::endl;
+    }
+
+    InitializeComputationOfMappingMatrix();
+    CreateSearchTreeWithAllNodesInOriginModelPart();
+    ComputeMappingMatrix();
+
+    KRATOS_INFO("ShapeOpt") << "Finished updating of mapper in " << timer.ElapsedSeconds() << " s." << std::endl;
+}
+
+void MapperVertexMorphingSymmetric::InitializeComputationOfMappingMatrix()
+{
+    mpSearchTree.reset();
+    mMappingMatrix.clear();
+}
+
+void MapperVertexMorphingSymmetric::CreateFilterFunction()
+{
+    const std::string filter_type = mMapperSettings["filter_function_type"].GetString();
+    const double filter_radius = mMapperSettings["filter_radius"].GetDouble();
+
+    mpFilterFunction = Kratos::make_unique<FilterFunction>(filter_type, filter_radius);
+}
+
+void MapperVertexMorphingSymmetric::InitializeMappingVariables()
+{
+    const unsigned int origin_node_number = mrOriginModelPart.Nodes().size();
+    const unsigned int destination_node_number = mrDestinationModelPart.Nodes().size();
+    mMappingMatrix.resize(destination_node_number*3,origin_node_number*3,false);
+}
+
+void MapperVertexMorphingSymmetric::AssignMappingIds()
+{
+    unsigned int i = 0;
+
+    // Note: loop in the same order as in AllocateMatrix(), to avoid reallocations of the matrix.
+    for(auto& node_i : mrOriginModelPart.Nodes())
+        node_i.SetValue(MAPPING_ID,i++);
+
+    i = 0;
+    for(auto& node_i : mrDestinationModelPart.Nodes())
+        node_i.SetValue(MAPPING_ID,i++);
+}
+
+void MapperVertexMorphingSymmetric::CreateSearchTreeWithAllNodesInOriginModelPart()
+{
+    mpSearchTree = Kratos::make_unique<KDTree>(mpSymmetry->GetOriginSearchNodes().begin(), mpSymmetry->GetOriginSearchNodes().end(), mBucketSize);
+}
+
+void MapperVertexMorphingSymmetric::ComputeMappingMatrix()
+{
+    AllocateMatrix();
+    const double filter_radius = mMapperSettings["filter_radius"].GetDouble();
+    const unsigned int max_number_of_neighbors = mMapperSettings["max_nodes_in_filter_radius"].GetInt();
+
+    #pragma omp parallel
+    {
+
+    // variable size, fixed capacity vecs
+    std::vector<bool> transform;
+    NodeVector total_neighbor_nodes;
+    std::vector<double> total_list_of_weights;
+    std::vector<double> list_of_weights;
+    transform.reserve(max_number_of_neighbors);
+    total_neighbor_nodes.reserve( max_number_of_neighbors );
+    total_list_of_weights.reserve( max_number_of_neighbors );
+    list_of_weights.reserve( max_number_of_neighbors );
+
+    // fixed size vec
+    NodeVector neighbor_nodes( max_number_of_neighbors );
+    #pragma omp for
+    for (int i = 0; i < int(mrDestinationModelPart.NumberOfNodes()); ++i)
+    {
```

Comment:
I think so too, but then you have to make many things threead_local

As it is right now you have plenty of race conditions in there, I am surprised the test don't fail

maybe it is better to make this parallel this in a different PR as it is a bit of work if you are not familiar with it

If you want I can point out the issues in more detail

Reply:
> If you want I can point out the issues in more detail

Yeah that yould actually be quite helpful, since i though i have considered everything.
Is it because i am calling non const functions?

Reply:
Since its been a while for me: Can sombebody explain me in brief what are the specifics of the following type of functions:

   - `IndexPartition<int>(mrOriginModelPart.NumberOfNodes()).for_each([&](const int Index) {...}`
   - `block_for_each(mrDestinationModelPart.Nodes(), [&](ModelPart::NodeType& rNode) {...}`

Why using them instead of "regular loops". I can guess it's because of parallelism and optimization reasons but would be nice if somebody with insight could explain it a bit.